WordPress/gutenberg
# Spacing controls: using CustomSelectControlV2 for >= 8 spacing sizes

## What?

Fixes: https://github.com/WordPress/gutenberg/issues/64270

Where there are more than a defined number of spacing sizes, the spacing input control component will render custom component, rather than the slider controls.

The current threshold for this behaviour is 8 spacing sizes:

https://github.com/WordPress/gutenberg/blob/75d0173916bee81173e99a9294f6d5bdc384f181/packages/block-editor/src/components/spacing-sizes-control/utils.js#L15

A theme.json that contains more than this number causes the browser to crash.


## Why?
CustomSelectControlV2Legacy was removed in favour of V2 but it appears the spacing input control component was not updated.

See: 


- https://github.com/WordPress/gutenberg/pull/63190 (uses legacy component)
- https://github.com/WordPress/gutenberg/pull/63258 (removed legacy component)

## How?

Exporting `CustomSelectControlV2` from components private APIs for use in spacing-input-control.js.

A slight refactor of the component's children was required.

## Testing Instructions

Things should work as they do before https://github.com/WordPress/gutenberg/pull/63258.

You can compare this PR and the previously working version by deactivating the Gutenberg plugin.

1. With a theme with 8+ values in settings.spacing.spacingSizes in theme.json
2. Create or edit a post
3. Add a group, or buttons, or anything with dimension controls
4. Click the "styles" tab in the sidebar
5. Check that you can add padding/margin/block gap sizes to the block and that they publish okay.
6. Also make sure you can toggle between custom/preset spacing sizes
7. Check in global styles and the site editor as well for good luck.

<details>
<summary>Example theme.json</summary>

```json
{
	"$schema": "../../schemas/json/theme.json",
	"version": 3,
	"settings": {
		"appearanceTools": true,
		"spacing": {
			"spacingSizes": [
				{
					"name": "1",
					"size": "1rem",
					"slug": "10"
				},
				{
					"name": "2",
					"size": "min(1.5rem, 2vw)",
					"slug": "20"
				},
				{
					"name": "3",
					"size": "min(2.5rem, 3vw)",
					"slug": "30"
				},
				{
					"name": "4",
					"size": "min(4rem, 5vw)",
					"slug": "40"
				},
				{
					"name": "5",
					"size": "min(6.5rem, 8vw)",
					"slug": "50"
				},
				{
					"name": "6",
					"size": "min(10.5rem, 13vw)",
					"slug": "60"
				},
				{
					"name": "7",
					"size": "min(10.5rem, 17vw)",
					"slug": "70"
				},
				{
					"name": "8",
					"size": "min(10.5rem, 21vw)",
					"slug": "80"
				}
			],
			"units": ["%", "px", "em", "rem", "vh", "vw"]
		}
	}
}

```
</details>



## Screenshots or screencast 



### With the plugin turned off

<img width="273" alt="Screenshot 2024-08-06 at 11 00 18 AM" src="https://github.com/user-attachments/assets/305aa1fe-5eab-4129-9826-bc94b0f52036">

### With the plugin turned on (trunk)

<img width="940" alt="Screenshot 2024-08-06 at 11 02 44 AM" src="https://github.com/user-attachments/assets/b40f8f3c-5a4b-4ec7-9fc6-34e8acff76c2">



### This PR with the plugin turned on

https://github.com/user-attachments/assets/4e94f584-9f58-46cd-8cb2-75c27987e18d
